File: install/extract_text_from_image.swift
#!/usr/bin/env swift

import Foundation
import Vision

func writeError(_ message: String) {
    let data = (message + "\n").data(using: .utf8)!
    FileHandle.standardError.write(data)
}

guard CommandLine.arguments.count == 2 else {
    writeError("Usage: extract_text_from_image.swift <image_path>")
    exit(1)
}

let imagePath = CommandLine.arguments[1]
guard FileManager.default.fileExists(atPath: imagePath) else {
    writeError("Error: Image file does not exist: \(imagePath)")
    exit(1)
}

guard let imageURL = URL(fileURLWithPath: imagePath) as URL?,
      let imageData = try? Data(contentsOf: imageURL) else {
    writeError("Error: Failed to load image")
    exit(1)
}

let requestHandler = VNImageRequestHandler(data: imageData, options: [:])
let textRequest = VNRecognizeTextRequest()
textRequest.recognitionLevel = .fast

// Configure recognition languages to support all available languages
// Supported languages as of macOS 13.0/iOS 16.0+:
// English, French, Italian, German, Spanish, Portuguese (Brazil),
// Chinese (Simplified/Traditional), Cantonese (Simplified/Traditional),
// Korean, Japanese, Russian, Ukrainian
// Note: To get the exact list at runtime, use:
// try textRequest.supportedRecognitionLanguages(for: .fast, revision: VNRecognizeTextRequestRevision1)
textRequest.recognitionLanguages = [
    "en-US",    // English
    "fr-FR",    // French
    "it-IT",    // Italian
    "de-DE",    // German
    "es-ES",    // Spanish
    "pt-BR",    // Portuguese
    "zh-Hans",  // Simplified Chinese
    "zh-Hant",  // Traditional Chinese
    "yue-Hans", // Simplified Cantonese
    "yue-Hant", // Traditional Cantonese
    "ko-KR",    // Korean
    "ja-JP",    // Japanese
    "ru-RU",    // Russian
    "uk-UA",    // Ukrainian
]

// Enable automatic language detection as fallback
// This helps when the text contains multiple languages or languages not in the list above
textRequest.automaticallyDetectsLanguage = true

do {
    try requestHandler.perform([textRequest])
} catch {
    writeError("Error: Vision framework request failed: \(error)")
    exit(1)
}

guard let observations = textRequest.results, !observations.isEmpty else {
    // No text found - exit with code 1 but no error message (this is expected)
    exit(1)
}

// Group observations by Y-coordinate to preserve line breaks
// Observations with similar Y-coordinates are on the same line
var lineGroups: [[VNRecognizedTextObservation]] = []
var currentLine: [VNRecognizedTextObservation] = []
var lastY: CGFloat? = nil
let yTolerance: CGFloat = 10.0 // Pixels - text within this Y range is considered same line

// Sort observations by Y-coordinate (top to bottom)
let sortedObservations = observations.sorted { obs1, obs2 in
    let y1 = obs1.boundingBox.midY
    let y2 = obs2.boundingBox.midY
    if abs(y1 - y2) < yTolerance {
        // Same line, sort by X (left to right)
        return obs1.boundingBox.minX < obs2.boundingBox.minX
    }
    return y1 < y2
}

for observation in sortedObservations {
    let currentY = observation.boundingBox.midY
    
    if let lastYValue = lastY {
        // Check if this observation is on a new line
        if abs(currentY - lastYValue) > yTolerance {
            // New line detected
            if !currentLine.isEmpty {
                lineGroups.append(currentLine)
                currentLine = []
            }
        }
    }
    
    currentLine.append(observation)
    lastY = currentY
}

// Add the last line
if !currentLine.isEmpty {
    lineGroups.append(currentLine)
}

// Extract text from each line group
var extractedLines: [String] = []
for lineGroup in lineGroups {
    var lineTextParts: [String] = []
    for observation in lineGroup {
        let topCandidates = observation.topCandidates(1)
        guard let topCandidate = topCandidates.first else {
            continue
        }
        lineTextParts.append(topCandidate.string)
    }
    if !lineTextParts.isEmpty {
        // Join words on the same line with spaces
        extractedLines.append(lineTextParts.joined(separator: " "))
    }
}

// Join lines with newlines to preserve line breaks
let extractedText = extractedLines.joined(separator: "\n")
if extractedText.isEmpty {
    exit(1)
}

print(extractedText)

File: install/extract_text_from_image.py
#!/usr/bin/env python3
"""
Extract text from an image using EasyOCR.
Similar to install/extract_text_from_image.swift for macOS, but uses EasyOCR instead of Vision framework.
"""

import sys
import os

def write_error(message: str) -> None:
    """Write error message to stderr."""
    sys.stderr.write(f"{message}\n")
    sys.stderr.flush()

def main() -> int:
    """Main function to extract text from image."""
    # Check command-line arguments
    if len(sys.argv) != 2:
        write_error("Usage: extract_text_from_image.py <image_path>")
        return 1
    
    image_path = sys.argv[1]
    
    # Verify image file exists
    if not os.path.exists(image_path):
        write_error(f"Error: Image file does not exist: {image_path}")
        return 1
    
    try:
        # Import EasyOCR
        import easyocr
    except ImportError:
        write_error("Error: easyocr module not found. Please install it with: pip install easyocr")
        return 1
    
    try:
        # Initialize EasyOCR reader
        # This will download models on first use (may be slow)
        # EasyOCR supports 80+ languages. Common language codes:
        # en (English), fr (French), de (German), es (Spanish), it (Italian),
        # pt (Portuguese), ch_sim (Simplified Chinese), ch_tra (Traditional Chinese),
        # ja (Japanese), ko (Korean), ru (Russian), uk (Ukrainian), th (Thai), etc.
        # See https://www.jaided.ai/easyocr/ for full list
        # gpu=False uses CPU only (faster install, no CUDA dependencies)
        languages = [
            'en',      # English
            'ch_tra',  # Traditional Chinese
        ]
        reader = easyocr.Reader(languages, gpu=False)
        
        # Read text from image
        results = reader.readtext(image_path)
        
        # Group text by Y-coordinate to preserve line breaks
        # results is a list of tuples: (bbox, text, confidence)
        # bbox is a list of 4 points: [[x1, y1], [x2, y2], [x3, y3], [x4, y4]]
        # We'll use the average Y-coordinate to group into lines
        y_tolerance = 10.0  # Pixels - text within this Y range is considered same line
        
        # Sort results by Y-coordinate (top to bottom)
        def get_y_center(bbox):
            """Get the center Y-coordinate of a bounding box."""
            y_coords = [point[1] for point in bbox]
            return sum(y_coords) / len(y_coords)
        
        def get_x_min(bbox):
            """Get the minimum X-coordinate for sorting within a line."""
            x_coords = [point[0] for point in bbox]
            return min(x_coords)
        
        sorted_results = sorted(results, key=lambda r: (get_y_center(r[0]), get_x_min(r[0])))
        
        # Group into lines
        line_groups = []
        current_line = []
        last_y = None
        
        for (bbox, text, confidence) in sorted_results:
            current_y = get_y_center(bbox)
            
            if last_y is not None:
                # Check if this text is on a new line
                if abs(current_y - last_y) > y_tolerance:
                    # New line detected
                    if current_line:
                        line_groups.append(current_line)
                        current_line = []
            
            current_line.append((bbox, text, confidence))
            last_y = current_y
        
        # Add the last line
        if current_line:
            line_groups.append(current_line)
        
        # Extract text from each line group
        extracted_lines = []
        for line_group in line_groups:
            line_text_parts = [text for (_, text, _) in line_group]
            if line_text_parts:
                # Join words on the same line with spaces
                extracted_lines.append(" ".join(line_text_parts))
        
        # Join lines with newlines to preserve line breaks
        extracted_text = "\n".join(extracted_lines)
        
        if not extracted_text.strip():
            # No text found - exit with code 1 but no error message (this is expected)
            return 1
        
        # Output extracted text to stdout
        print(extracted_text)
        return 0
        
    except Exception as e:
        write_error(f"Error: EasyOCR text extraction failed: {e}")
        return 1

if __name__ == "__main__":
    sys.exit(main())
